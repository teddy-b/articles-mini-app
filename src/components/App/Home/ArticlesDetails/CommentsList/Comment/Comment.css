.comment {
  padding: 1em 0;
}

.comment .comment-heading {
  display: flex;
  justify-content: space-between;
}

.comment .comment-author {
 font-weight: bold;
}

.comment .comment-created-at {
  font-style: italic;
}

.comment .comment-reply-active {
  cursor: pointer;
  color: blue;
  text-decoration: underline;
}
