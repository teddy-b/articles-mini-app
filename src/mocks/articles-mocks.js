export const articlesMock = [
  {
    title: 'CERN experiment discovers five new particles',
    text: 'A new group of particles which have long been hiding in plain sight have finally been discovered thanks to the incredibly sensitive LHCb experiment at CERN. LHCb, otherwise known as the Large Hadron Collider beauty experiment, is one of seven experiments collecting data at the Large Hadron Collider at CERN. UK participation in the experiment is funded by STFC.',
    imageUrl: 'http://www.stfc.ac.uk/stfc/cache/file/4821074D-0D31-4964-9318CA0AE81E0639.jpg',
    id: 0,
    createdAt: 1530698004815,
    commentsCount: 8
  },
  {
    title: 'Show HN: Marta – A fast and minimalistic file manager for macOS',
    text: 'File Manager for macOS. Native. Minimalistic. Fast. ',
    imageUrl: 'https://marta.yanex.org/data/images/dark.png',
    id: 1,
    createdAt: 1530698004815,
    commentsCount: 7
  },
  {
    title: 'Tears',
    text: 'So humans are a peculiar species. You probably kne…specific traits and behaviors are uniquely human.',
    imageUrl: 'http://www.meltingasphalt.com/wp-content/uploads/2014/12/weeping_equation.png',
    id: 2,
    createdAt: 1530698004815,
    commentsCount: 0
  }
];

export default articlesMock;
