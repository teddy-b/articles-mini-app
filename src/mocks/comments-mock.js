export const commentsMock = [
  {
    articleId: 0,
    author: 'grabcocque',
    createdAt: 1530698004815,
    id: 0,
    parentCommentId: null,
    repliesCount: 0,
    text: 'I have to admit, I rather admire their putting a brave face on things, with their mildly-too-insistent claim that the LHC is vital for finding new physics despite the increasing likelihood it\'s not going to find any. In that sense the LHC is a political failure, because it has failed at its primary political job, which is to make the argument for an even bigger collider.'
  },
  {
    articleId: 0,
    author: 'teddythetwig',
    createdAt: 1530698004815,
    id: 1,
    parentCommentId: null,
    repliesCount: 2,
    text: 'I assume these particles were already predicted as part of the Standard Model?'
  },
  {
    articleId: 0,
    author: 'guitarbill',
    createdAt: 1530698004815,
    id: 6,
    parentCommentId: null,
    repliesCount: 0,
    text: 'AFAIK, the experimentally determined rest masses of these excited states/particles (same thing, different was of looking at it) agree with the calculated ones well. So yes, they were predicted. No surprises sadly; of course it\'s still a huge achievement!'
  }
];

export default commentsMock;
