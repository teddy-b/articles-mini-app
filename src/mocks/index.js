export * from './articles-mocks';
export * from './comments-mock';
export * from './replies-mock';
export * from './replies-replies-mock';
