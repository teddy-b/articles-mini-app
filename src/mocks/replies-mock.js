export const repliesMock = [
  {
    articleId: 0,
    author: 'brutuscat',
    createdAt: 1530698004815,
    id: 2,
    parentCommentId: 1,
    repliesCount: 0,
    text: 'Professor Tim Gershon, Professor of Physics at University of Warwick and UK spokesperson for the LHCb experiment: '
  },
  {
    articleId: 0,
    author: 'gpderetta',
    createdAt: 1530698004815,
    id: 3,
    parentCommentId: 1,
    repliesCount: 2,
    text: 'yes, and note that these are not fundamental particles (like the Higgs was for example), but composite particles of yet another combination of the fundamental quarks. The SM predicts the existence of hundreds (thousands?) of these.'
  }
];

export default repliesMock;
