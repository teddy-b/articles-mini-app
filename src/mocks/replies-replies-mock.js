export const repliesRepliesMock = [
  {
    articleId: 0,
    author: 'inlineint',
    createdAt: 1530698004815,
    id: 4,
    parentCommentId: 3,
    repliesCount: 0,
    text: 'Actually it is even a one new particle (Omega_c baryon), but they observed five different excited energy states of it (like observing different excited states of a Hydrogen atom), so called resonances [1]. But the discovery is still exciting because it should have been really hard to find something that we don\'t really know how looks like in such large amount of noise.'
  },
  {
    articleId: 0,
    author: 'binarymax',
    createdAt: 1530698004815,
    id: 5,
    parentCommentId: 3,
    repliesCount: 0,
    text: 'Because a prediction is just an assumption (theory), and it can become a house of cards when basing future science on that assumption. Observation is proof, so future science can use that proof without worry.'
  }
];

export default repliesRepliesMock;
