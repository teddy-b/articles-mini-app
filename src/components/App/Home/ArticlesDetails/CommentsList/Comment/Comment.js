import React, { Component } from 'react';
import moment from 'moment';

import RepliesList from '../RepliesList';

import './Comment.css';

class Comment extends Component {
  constructor(props) {
    super(props);
    this.state = {
      replyVisible: false
    };
  }

  render() {
    const {
      author,
      createdAt,
      id,
      onAddReply,
      onLoadReplies,
      replies,
      repliesCount,
      text
    } = this.props;
    const { replyVisible } = this.state;
    return (
      <div className="comment">
        <div className="comment-heading">
          <span className="comment-author">{author}</span>
          <span className="comment-created-at">{moment(createdAt).fromNow()}</span>
          <span
            className={repliesCount && 'comment-reply-active comment-reply-active-count'}
            onClick={() => repliesCount && onLoadReplies(id)}
          >
            {repliesCount} replies
          </span>
          {!replyVisible && (
            <span
              className="comment-reply-active comment-reply-active-reply"
              onClick={() => this.setState({
                replyVisible: true
              })}
            >
              reply
            </span>
          )}
          {replyVisible && (
            <span
              className="comment-reply-active comment-reply-active-discard"
              onClick={() => this.setState({
                replyVisible: false
              })}
            >
              discard reply
            </span>
          )}
        </div>
        <p>{text}</p>
        {replyVisible && (
          <textarea
            rows="5"
            cols="40"
            placeholder="Add reply"
            onKeyPress={(e) => {
              if (e.key === 'Enter') {
                onAddReply(id, e.target.value);
                e.target.value = '';
                e.target.blur();
                this.setState({
                  replyVisible: false
                });
              }
            }}
          />
        )}
        {replies && (
          <RepliesList
            onAddReply={onAddReply}
            onLoadReplies={onLoadReplies}
            replies={replies}
          />
        )}
      </div>
    );
  }
}

export default Comment;
