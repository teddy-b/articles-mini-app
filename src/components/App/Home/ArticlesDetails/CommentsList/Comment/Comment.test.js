import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { repliesMock } from '../../../../../../mocks';

import Comment from './Comment';

jest.mock('../RepliesList', () => () => 'RepliesList');

describe('Comment', () => {
  const onAddReplyMock = jest.fn();
  const onLoadReplieMock = jest.fn();
  const props = {
    ...repliesMock[0],
    createdAt: Date.now(),
    onAddReply: onAddReplyMock,
    onLoadReplies: onLoadReplieMock,
    replies: []
  };

  it('renders correctly', () => {
    const component = renderer.create(
      <Comment
        {...props}
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('add reply on Enter', () => {
    const component = mount(
      <Comment
        {...props}
      />
    );

    component.find('.comment-reply-active-reply').simulate('click');

    component.find('textarea').simulate('keypress', {
      key: 'a'
    });

    expect(onAddReplyMock).not.toHaveBeenCalled();

    component.find('textarea').simulate('keypress', {
      key: 'Enter'
    });

    expect(onAddReplyMock).toHaveBeenCalled();
  });

  it('load replies on click', () => {
    const component = mount(
      <Comment
        {...props}
        replies={repliesMock}
        repliesCount={repliesMock.length}
      />
    );

    component.find('.comment-reply-active-count').simulate('click');

    expect(onLoadReplieMock).toHaveBeenCalled();
  });

  it('show input on click', () => {
    const component = mount(
      <Comment
        {...props}
        replies={repliesMock}
        repliesCount={repliesMock.length}
      />
    );

    component.find('.comment-reply-active-reply').simulate('click');

    expect(component.find('textarea').exists()).toBe(true);
  });

  it('hide input on click', () => {
    const component = mount(
      <Comment
        {...props}
        replies={repliesMock}
        repliesCount={repliesMock.length}
      />
    );

    component.find('.comment-reply-active-reply').simulate('click');

    expect(component.find('textarea').exists()).toBe(true);

    component.find('.comment-reply-active-discard').simulate('click');

    expect(component.find('textarea').exists()).toBe(false);
  });
});
