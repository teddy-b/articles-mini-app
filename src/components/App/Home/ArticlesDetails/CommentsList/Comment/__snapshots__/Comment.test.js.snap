// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment renders correctly 1`] = `
<div
  className="comment"
>
  <div
    className="comment-heading"
  >
    <span
      className="comment-author"
    >
      brutuscat
    </span>
    <span
      className="comment-created-at"
    >
      a few seconds ago
    </span>
    <span
      className={0}
      onClick={[Function]}
    >
      0
       replies
    </span>
    <span
      className="comment-reply-active comment-reply-active-reply"
      onClick={[Function]}
    >
      reply
    </span>
  </div>
  <p>
    Professor Tim Gershon, Professor of Physics at University of Warwick and UK spokesperson for the LHCb experiment: 
  </p>
  RepliesList
</div>
`;
